README
NAME
    MooX::Traits - automatically apply roles at object creation time

SYNOPSIS
    Given some roles:

       package Role;
       use Moo::Role;
       has foo => ( is => 'ro', required => 1 );

    And a class:

       package Class;
       use Moo;
       with 'MooX::Traits';

    Apply the roles to the class at new time:

       my $object = Class->with_traits('Role')->new( foo => 42 );

    Then use your customized class:

       $object->isa('Class'); # true
       $object->does('Role'); # true
       $object->foo; # 42

DESCRIPTION
    Was any of the SYNOPSIS unexpected? Basically, this module is the same
    thing as MooseX::Traits and MouseX::Traits, only for Moo. *Quelle
    surprise*, right?

  Methods
    `$class->with_traits( @traits )`
        Return a new class name with the traits applied.

    `$class->new_with_traits(%args, traits => \@traits)`
        `new_with_traits` can also take a hashref, e.g.:

           my $instance = $class->new_with_traits({ traits => \@traits, foo => 'bar' });

        This method exists for compatibility with the MooseX and MouseX
        equivalents of this module, but generally speaking you should prefer
        to use `with_traits`.

    `$class->_trait_namespace`
        This returns undef, but you can override it in your class to
        automatically prepend a namespace to supplied traits.

        This differs slightly from the MooseX and MouseX versions of this
        module which have `_trait_namespace` as an attribute instead of a
        method.

  Compatibility
    Although called MooX::Traits, this module actually uses Role::Tiny, so
    doesn't really require Moo. If you use it in a non-Moo class, you should
    be able to safely consume any Role::Tiny-based traits.

    If you use it in a Moo class, you should also be able to consume
    Moo::Role-based traits and Moose::Role-based traits.

    Package::Variant and MooseX::Role::Parameterized roles should be usable;
    just provide a hashref of arguments:

       $class->with_traits( $param_role => \%args )->new( ... )

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooX-Traits>.

SEE ALSO
    Moo::Role, Role::Tiny.

    MooseX::Traits, MouseX::Traits.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooX-Traits
===========

Created:      2014-05-24
Home page:    <https://metacpan.org/release/MooX-Traits>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooX-Traits>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2014-05-29

 [ Bug Fixes ]
 - MooX::Traits::Util should also load UNIVERSAL::DOES.

0.003	2014-05-29

 [ Bug Fixes ]
 - Load UNIVERSAL::DOES on Perls older than 5.10.

0.002	2014-05-29

 - Added: Bundle a MooX::Traits::Util module, broadly similar to
   MooseX::Traits::Util.
 - Added: Provide a `new_class_with_traits_one_by_one` function.
   Torsten Raudssus++

0.001	2014-05-25	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "automatically apply roles at object creation time",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 1,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.016, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "MooX-Traits",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt::Profile::TOBYINK" => 0.001 } },
                        runtime   => {
                                       recommends => { "Moo::Role" => 0 },
                                       requires   => {
                                                       "Exporter::Shiny" => 0,
                                                       "Module::Runtime" => 0,
                                                       "perl" => "5.006000",
                                                       "Role::Tiny" => "1.000000",
                                                       "Scalar::Util" => 0,
                                                     },
                                     },
                        test      => {
                                       recommends => {
                                                       "Moo" => 0,
                                                       "MooseX::Role::Parameterized" => 0,
                                                       "Package::Variant" => 0,
                                                       "Test::Fatal" => 0,
                                                     },
                                       requires   => { "Test::More" => 0.96, "Test::Requires" => 0 },
                                     },
                      },
  "provides"       => {
                        "MooX::Traits"       => { file => "lib/MooX/Traits.pm", version => 0.004 },
                        "MooX::Traits::Util" => { file => "lib/MooX/Traits/Util.pm", version => 0.004 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=MooX-Traits" },
                        homepage     => "https://metacpan.org/release/MooX-Traits",
                        license      => ["http://dev.perl.org/licenses/"],
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/MooX-Traits/project",
                      },
  "version"        => 0.004,
};

my %dynamic_config;
do {
$meta->{prereqs}{runtime}{requires}{'UNIVERSAL::DOES'} = 0 if $] < 5.010000;

};

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t t/from-moosex-traits/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

